Various additional fields for ATSD AGOL view

mddilley opened the pull request:
## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/17606

This PR adds four additional columns to the `component_arcgis_online_view` DB view:
- `council_districts_searchable` (makes council districts more searchable in ArcGIS)
- `substantial_completion_date_estimated` (see explanation in issue and test steps below)
- `related_project_ids` (shows all subproject ids in a text list)
- `related_project_ids_searchable` (shows all subproject ids in a list that is more searchable in ArcGIS)

## Testing
**URL to test:** 
<!---
 [branch-name]--atd-moped-main.netlify.app/moped/ if test instance or deploy-preview-[pr-number]--atd-moped-main.netlify.app/moped/ 
--->
Local

**Steps to test:**
1. Start up the local stack with production data. Create a new project and add a component that spans multiple council disctricts. I created a bike lane that goes east to west from Neches to San Marcos on 7th St. Note the component id in the url for future DB queries.
2. Open up your database client and run the following query:
```sql
SELECT
	council_districts,
	council_districts_searchable
FROM
	component_arcgis_online_view
WHERE
	project_component_id = <your component id>;
```
3. You should see that single digit districts have one zero padding to the left in the new searchable version of the column.
4. Now, add three subprojects to your project in the project summary view. Then, test the following query:
```sql
SELECT
	related_project_ids,
	related_project_ids_searchable
FROM
	component_arcgis_online_view
WHERE
	project_component_id = <your component id>;
```
5. You should see a list of the subproject ids in related project ids column and a list of related project ids with left-padded zeroes going up to 5 digits in the searchable columns. These lists should include the project you created (parent of the child projects).
6. Next, delete all phases from your project in the **Timeline** tab. Add a phase that is either **Complete** or **Post-construction** with a start date that is before today with the **Confirmed** checkbox unchecked. Save.
7. Check the `substantial_completion_date_estimated` with the following query:
```sql
SELECT
	substantial_completion_date, substantial_completion_date_estimated
FROM
	component_arcgis_online_view
WHERE
	project_component_id = <your component id>
```
8. The `substantial_completion_date_estimated` should contain the date of your new phase. The `substantial_completion_date` should be null.
9. Add another phase that is either **Complete** or **Post-construction** with a end date earlier than your last phase and the **Confirmed** checkbox unchecked. The `substantial_completion_date_estimated` should have the date of your earliest phase start or end date and `substantial_completion_date` should still be null.
10. Add another phase that is either **Complete** or **Post-construction** with a start or end date before today and leave the **Confirmed** checkbox checked. Now, `substantial_completion_date` should have your confirmed date and `substantial_completion_date_estimated` should be null. 


---
#### Ship list
- [x] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

frankhereford reviewed (approved):
Works great and the code looks 💯! Thanks Mike!

johnclary commented on moped-database/views/component_arcgis_online_view.sql:
```diff
@@ -13,7 +13,8 @@ CREATE OR REPLACE VIEW component_arcgis_online_view AS WITH work_types AS (
         features.component_id AS project_component_id,
-        string_agg(DISTINCT features_council_districts.council_district_id::text, ', '::text) AS council_districts
+        string_agg(DISTINCT features_council_districts.council_district_id::text, ', '::text) AS council_districts,
+        string_agg(DISTINCT lpad(features_council_districts.council_district_id::text, 2, '0'::text), ', '::text) AS council_districts_searchable
```
TIL. postgres rules!

mddilley replied:
yeah - `lpad` was new to me. they've thought of everything!

johnclary commented on moped-database/migrations/1718314926409_add_council_districts_searchable/up.sql:
```diff
@@ -0,0 +1,235 @@
+        least(
+            min_dates.min_phase_start,
+            min_dates.min_phase_end
+        ) AS min_phase_date
```
Mike, this SQL looks fantastic. one question—do you think this should be moved up to the project list view to sit alongside the `substantial_completion_date` we keep there? it might be good to have for if/when we want to expose it through the Moped editor UI.

mddilley replied:
i thought about this too and went with the easiest route, but you make a good case for having the two substantial fields equally accessible for future needs. I need to make another update in here before merge so I'll pull this up into the plv!

johnclary reviewed (approved):
I ran through the test steps and also did a bit of benchmarking—this looks excellent. 

I have a question about moving the new date column to the project list view, but this is good to go as-is. Thanks, Mike! 🚢

chiaberry reviewed (approved):
test steps check out! 🚢

mddilley commented:
y'all - i made a couple changes here:
- i added a filter so that the related project ids lists would not contain soft deleted projects (this was feedback from Nathan)
- i pulled the `substantial_completion_date_estimated` up into the `project_list_view` and refactored the way `substantial_completion_date` works in the project list view since I couldn't use a case statement on a column calculated through a subquery (pg error column doesn't exist).
   - i tested the results from the current way `substantial_completion_date` it is calculated against the new way in the last commit and 0 rows had a different result

I might re-request reviews tomorrow, and this probably won't make the release anyway since the renamed fields haven't been updated in the geo dataset yet. I'll keep you all updated.

mddilley commented on moped-database/views/component_arcgis_online_view.sql:
```diff
@@ -100,6 +101,35 @@ component_tags AS (
+        concat_ws(', '::text, lpad(pmp.project_id::text, 5, '0'::text), string_agg(lpad(cmp.project_id::text, 5, '0'::text), ', '::text)) AS related_project_ids_searchable_with_self
+    FROM moped_project pmp
+    LEFT JOIN moped_project cmp ON pmp.project_id = cmp.parent_project_id
+    WHERE cmp.is_deleted = false
```
🔖 this is new

mddilley commented on moped-database/views/component_arcgis_online_view.sql:
```diff
@@ -126,7 +156,7 @@ SELECT
     mpc.completion_date,
     coalesce(mpc.completion_date, plv.substantial_completion_date) AS substantial_completion_date,
-    '2024-01-01T00:00:00-06:00'::text AS substantial_completion_date_estimated,
+    plv.substantial_completion_date_estimated,
```
🔖 this is new too

mddilley commented on moped-database/views/project_list_view.sql:
```diff
@@ -95,6 +95,58 @@ parent_child_project_map AS (
     GROUP BY projects.project_id
+),
+
+min_confirmed_phase_dates AS (
```
🔖 from here down to 219 is new and updates the way the project list view get the `substantial_completion_date` and then uses it in a case statement to get `substantial_completion_date_estimated`

johnclary reviewed (approved):
looks great! thanks for reworking `substantial_completion_date` to tighten everything up. ship it! 🚢
